File: src/detection/npm.ts
import type { InstallDetection, Evidence } from '../types.js';
import { execFile } from 'node:child_process';
import { promisify } from 'node:util';
import { lstat, readlink } from 'node:fs/promises';

const execFileAsync = promisify(execFile);

/** npm global install path patterns */
const NPM_PATH_PATTERNS = [
  'node_modules/.bin/',
  '/lib/node_modules/',
];

/**
 * Detect install channel from npm global install indicators.
 * Checks path patterns, npm global prefix, and symlink targets.
 * Confidence is based on the number of matching evidence items.
 */
export async function detectFromNpm(
  execPath: string,
): Promise<InstallDetection | null> {
  const evidence: Evidence[] = [];

  // 1. Path pattern matching
  const matchedPattern = NPM_PATH_PATTERNS.find((pattern) =>
    execPath.includes(pattern),
  );

  if (matchedPattern) {
    evidence.push({
      source: 'path_pattern',
      detail: `Exec path matches npm pattern: ${matchedPattern}`,
    });
  }

  // 2. npm prefix -g comparison
  try {
    const { stdout } = await execFileAsync('npm', ['prefix', '-g']);
    const globalPrefix = stdout.trim();
    if (execPath.startsWith(globalPrefix)) {
      evidence.push({
        source: 'npm_prefix',
        detail: `Exec path is under npm global prefix: ${globalPrefix}`,
      });
    }
  } catch {
    // npm command failed — ignore
  }

  // 3. Symlink check (npm link installs)
  try {
    const stats = await lstat(execPath);
    if (stats.isSymbolicLink()) {
      const target = await readlink(execPath);
      if (target.includes('node_modules')) {
        evidence.push({
          source: 'symlink',
          detail: `Symlink target includes node_modules: ${target}`,
        });
      }
    }
  } catch {
    // Symlink check failed — ignore
  }

  if (evidence.length === 0) {
    return null;
  }

  const confidence = evidence.length >= 2 ? 'high' : 'medium';

  return {
    channel: 'npm-global',
    confidence,
    evidence,
  };
}

File: rust/crates/update-kit/src/detection/npm.rs
use crate::types::{Channel, Confidence, Evidence, InstallDetection};
use crate::utils::process::CommandRunner;

/// Path patterns that indicate an npm global installation.
const NPM_PATH_PATTERNS: &[&str] = &["node_modules", "/lib/node_modules/", "/.npm/"];

/// Detects installation via npm global by checking path patterns and optionally
/// verifying with `npm prefix -g`.
///
/// If the executable path contains an npm pattern, detection is triggered.
/// Runs `npm prefix -g` to verify the path starts with the npm global prefix,
/// yielding High confidence on match. Otherwise, Medium confidence is returned.
pub async fn detect_from_npm(exec_path: &str, cmd: &dyn CommandRunner) -> Option<InstallDetection> {
    let matching_pattern = NPM_PATH_PATTERNS
        .iter()
        .find(|pattern| exec_path.contains(*pattern));

    let pattern = matching_pattern?;
    let mut evidence = vec![Evidence {
        source: "npm-path".into(),
        detail: format!("path contains npm pattern '{}'", pattern),
    }];

    // Try to verify with npm prefix -g
    match cmd.run("npm", &["prefix", "-g"]).await {
        Ok(output) if output.success() => {
            let prefix = output.stdout.trim().to_string();
            if exec_path.starts_with(&prefix) {
                evidence.push(Evidence {
                    source: "npm-verify".into(),
                    detail: format!("path starts with npm global prefix '{}'", prefix),
                });
                return Some(InstallDetection {
                    channel: Channel::NpmGlobal,
                    confidence: Confidence::High,
                    evidence,
                });
            }
        }
        _ => {
            // npm not available or failed, fall through to Medium
        }
    }

    Some(InstallDetection {
        channel: Channel::NpmGlobal,
        confidence: Confidence::Medium,
        evidence,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::utils::process::TokioCommandRunner;

    #[tokio::test]
    async fn npm_path_with_node_modules() {
        let cmd = TokioCommandRunner;
        let result = detect_from_npm("/usr/local/lib/node_modules/.bin/my-app", &cmd).await;
        assert!(result.is_some());
        let detection = result.unwrap();
        assert_eq!(detection.channel, Channel::NpmGlobal);
        // Confidence depends on whether npm is available in test env
        assert!(detection.confidence >= Confidence::Medium);
    }

    #[tokio::test]
    async fn non_npm_path_returns_none() {
        let cmd = TokioCommandRunner;
        let result = detect_from_npm("/usr/local/bin/my-app", &cmd).await;
        assert!(result.is_none());
    }

    #[tokio::test]
    async fn npm_dot_npm_path() {
        let cmd = TokioCommandRunner;
        let result = detect_from_npm("/home/user/.npm/bin/my-app", &cmd).await;
        assert!(result.is_some());
        let detection = result.unwrap();
        assert_eq!(detection.channel, Channel::NpmGlobal);
    }
}
